etcdctlv3: add timeout support

mqliang opened the pull request:
ref https://github.com/coreos/etcd/issues/4769

@xiang90 
Do not merge, just send for review, only adding support to "etcdctl member". If this approach looks good, will update to support all etcdctlv3 commands.

xiang90 commented on etcdctlv3/command/util.go:
```diff
@@ -47,3 +50,9 @@ func argify(s string) []string {
 	return r.FindAllString(s, -1)
 }
+
+func contextWithTotalTimeout(cmd *cobra.Command) (context.Context, context.CancelFunc) {
```
rename this to commandCtx()

xiang90 commented on etcdctlv3/command/util.go:
```diff
@@ -47,3 +50,9 @@ func argify(s string) []string {
+
+func contextWithTotalTimeout(cmd *cobra.Command) (context.Context, context.CancelFunc) {
+	var timeOut time.Duration
+	cmd.Flags().DurationVar(&timeOut, "total-timeout", 0, "")
```
this is not total-timeout. this is request timeout. we probably need another dial timeout for dial the connection (we dial in newClient func in global.go). yes, we can make a total-timeout but this is not.

xiang90 replied:
set a default timeout to something reasonable. probably 5 seconds?

mqliang replied:
"snaphot" and "txn" command seems time-consuming, is 5 seconds safe enough? Not familiar with "snaphot" and "txn" command though, just my intuition.

xiang90 replied:
txn is low cost. Snapshot involves `watch` so we shall not add timeout to it. basically if a cmd involves watch, we need to be careful.

xiang90 commented:
This approach looks good to me. We need to be careful with watch. watch is designed to live long. So we should not set timeout for watch by default.

mqliang commented:
@xiang90 Updated. PTAL

xiang90 commented on etcdctlv3/command/lease_command.go:
```diff
@@ -123,7 +126,9 @@ func leaseKeepAliveCommandFunc(cmd *cobra.Command, args []string) {
 	}
 
-	respc, kerr := mustClientFromCmd(cmd).KeepAlive(context.TODO(), v3.LeaseID(id))
+	ctx, cancel := commandCtx(cmd)
```
keep alive is a long running one. so we probably should not set timeout here.

xiang90 commented on etcdctlv3/command/make_mirror_command.go:
```diff
@@ -60,7 +60,9 @@ func makeMirrorCommandFunc(cmd *cobra.Command, args []string) {
 	c := mustClientFromCmd(cmd)
 
-	err := makeMirror(context.TODO(), c, dc)
+	ctx, cancel := commandCtx(cmd)
```
make mirror is also a long running command. we should not set the timeout for it.

mqliang replied:
Does "make mirror" means make a mirror of the whole cluster to elsewhere as backup?

xiang90 replied:
read https://github.com/coreos/etcd/blob/master/etcdctlv3/doc/mirror_maker.md?

xiang90 commented on etcdctlv3/command/util.go:
```diff
@@ -47,3 +50,9 @@ func argify(s string) []string {
+
+func commandCtx(cmd *cobra.Command) (context.Context, context.CancelFunc) {
+	var timeOut time.Duration
+	cmd.Flags().DurationVar(&timeOut, "timeout", 5*time.Second, "Timeout for the request")
```
make 5 second a const or a global var at least?

mqliang commented:
@xiang9 Fixed. PTAL

xiang90 commented on etcdctlv3/command/util.go:
```diff
@@ -18,10 +18,15 @@ import (
 	pb "github.com/coreos/etcd/storage/storagepb"
 )
 
+const defaultTimeOut = 5 * time.Second
```
defaultCommandTimeout?

xiang90 commented:
@mqliang Have you tried this flag? Does it work? I think you need to define this flag like https://github.com/coreos/etcd/blob/master/etcdctlv3/main.go#L51?

xiang90 commented on etcdctlv3/command/util.go:
```diff
@@ -47,3 +52,9 @@ func argify(s string) []string {
+
+func commandCtx(cmd *cobra.Command) (context.Context, context.CancelFunc) {
+	var timeOut time.Duration
+	cmd.Flags().DurationVar(&timeOut, "timeout", defaultTimeOut, "Timeout for the request")
```
i do not think this works. 

also probably we should call this `--command-timeout`, "Timeout for short running command (excluding dial timeout)"

mqliang commented:
Fail to build up a develop environment, could you please kindly help take a look?
```
root@commonqa-lmq-test-etcd:~/mqliang/src/github.com/coreos/etcd/etcdctlv3# curl https://discovery.etcd.io/new?size=1
https://discovery.etcd.io/6bdfe3420363168c997269c55391c026
```

```
root@commonqa-lmq-test-etcd:~/mqliang/src/github.com/coreos/etcd# etcd --name infra0 --initial-advertise-peer-urls http://10.180.157.55:2380 \
>   --listen-peer-urls http://10.180.157.55:2380 \
>   --listen-client-urls http://10.180.157.55:2379,http://127.0.0.1:2379 \
>   --advertise-client-urls http://10.180.157.55:2379 \
>   --discovery https://discovery.etcd.io/6bdfe3420363168c997269c55391c026
2016/03/22 16:38:31 no data-dir provided, using default data-dir ./infra0.etcd
2016/03/22 16:38:31 etcd: listening for peers on http://10.180.157.55:2380
2016/03/22 16:38:31 etcd: listening for client requests on http://10.180.157.55:2379
2016/03/22 16:38:31 etcd: listening for client requests on http://127.0.0.1:2379
2016/03/22 16:38:36 discovery: during cluster status check connection to https://discovery.etcd.io timed out, retrying in 2s
2016/03/22 16:38:44 discovery: found self 39f3acf28cac14e2 in the cluster
2016/03/22 16:38:44 discovery: found 1 needed peer(s)
2016/03/22 16:38:44 etcdserver: name = infra0
2016/03/22 16:38:44 etcdserver: data dir = infra0.etcd
2016/03/22 16:38:44 etcdserver: heartbeat = 100ms
2016/03/22 16:38:44 etcdserver: election = 1000ms
2016/03/22 16:38:44 etcdserver: snapshot count = 10000
2016/03/22 16:38:44 etcdserver: discovery URL= https://discovery.etcd.io/6bdfe3420363168c997269c55391c026
2016/03/22 16:38:44 etcdserver: advertise client URLs = http://10.180.157.55:2379
2016/03/22 16:38:44 etcdserver: initial advertise peer URLs = http://10.180.157.55:2380
2016/03/22 16:38:44 etcdserver: initial cluster = infra0=http://10.180.157.55:2380
2016/03/22 16:38:44 etcdserver: start member 39f3acf28cac14e2 in cluster 8643517ca5c4e4e7
2016/03/22 16:38:44 raft: 39f3acf28cac14e2 became follower at term 0
2016/03/22 16:38:44 raft: newRaft 39f3acf28cac14e2 [peers: [], term: 0, commit: 0, applied: 0, lastindex: 0, lastterm: 0]
2016/03/22 16:38:44 raft: 39f3acf28cac14e2 became follower at term 1
2016/03/22 16:38:44 etcdserver: added local member 39f3acf28cac14e2 [http://10.180.157.55:2380] to cluster 8643517ca5c4e4e7
2016/03/22 16:38:45 raft: 39f3acf28cac14e2 is starting a new election at term 1
2016/03/22 16:38:45 raft: 39f3acf28cac14e2 became candidate at term 2
2016/03/22 16:38:45 raft: 39f3acf28cac14e2 received vote from 39f3acf28cac14e2 at term 2
2016/03/22 16:38:45 raft: 39f3acf28cac14e2 became leader at term 2
2016/03/22 16:38:45 raft.node: 39f3acf28cac14e2 elected leader 39f3acf28cac14e2 at term 2
2016/03/22 16:38:45 etcdserver: published {Name:infra0 ClientURLs:[http://10.180.157.55:2379]} to cluster 8643517ca5c4e4e7
```


```
root@commonqa-lmq-test-etcd:~/mqliang/src/github.com/coreos/etcd/etcdctlv3# ./etcdctlv3 --endpoints=[10.180.157.55:2379] member list
2016/03/22 16:41:32 transport: http2Client.notifyError got notified that the client transport was broken unexpected EOF.
2016/03/22 16:41:33 grpc: Conn.resetTransport failed to create client transport: connection error: desc = "transport: write tcp 10.180.157.55:32962->10.180.157.55:2379: write: broken pipe"; Reconnecting to "10.180.157.55:2379"
2016/03/22 16:41:34 grpc: Conn.resetTransport failed to create client transport: connection error: desc = "transport: write tcp 10.180.157.55:32966->10.180.157.55:2379: write: broken pipe"; Reconnecting to "10.180.157.55:2379"
Error:  rpc error: code = 13 desc = "transport is closing"
```

xiang90 commented:
Version of etcd?

xiang90 commented:
the simplest way to bring up a v3 cluster is to run
```
go get github.com/coreos/etcd
cd $GOPATH/src/github.com/coreos/etcd
./build
go build -o bin/etcdctlv3 ./etcdctlv3
goreman start -f V3DemoProcfile start
./bin/etcdctlv3 put hello "v3 API"
```

mqliang commented:
@xiang90  Using the method you suggested to bring up a local cluster, server side and etcdctlv2 works fine. But etcdctlv3 doesn't work:
```
commit c7a7dfdc9aa2082d15439e719c88d7023a9dde88
Merge: b56c2e0 d883d5a
Author: Xiang Li <xiangli.cs@gmail.com>
Date:   Tue Mar 15 16:48:12 2016 -0700

    Merge pull request #4772 from xiang90/bk
    
    storage: support seq put to make db more compact
```
```
➜  etcd git:(master) ✗ ./build 
./build: line 24: [: 6rc2: integer expression expected
# github.com/coreos/etcd
link: warning: option -X github.com/coreos/etcd/version.GitSHA c7a7dfd may not work in future releases; use -X github.com/coreos/etcd/version.GitSHA=c7a7dfd
```

```
➜  etcd git:(master) ✗ go build -o bin/etcdctlv3 ./etcdctlv3
➜  etcd git:(master) ✗ ls ./bin/                            
etcd      etcdctl   etcdctlv3
```

```
➜  etcd git:(master) ✗ goreman start
13:51:09 etcd3 | Starting etcd3 on port 5002
13:51:09 etcd2 | Starting etcd2 on port 5001
13:51:09 proxy | Starting proxy on port 5003
13:51:09 etcd1 | Starting etcd1 on port 5000
13:51:09 etcd2 | 2016-03-23 13:51:09.746290 I | etcdmain: etcd Version: 2.3.0-alpha.1+git
13:51:09 etcd2 | 2016-03-23 13:51:09.746419 I | etcdmain: Git SHA: c7a7dfd
13:51:09 etcd2 | 2016-03-23 13:51:09.746425 I | etcdmain: Go Version: go1.6rc2
13:51:09 etcd2 | 2016-03-23 13:51:09.746431 I | etcdmain: Go OS/Arch: darwin/amd64
13:51:09 etcd2 | 2016-03-23 13:51:09.746442 I | etcdmain: setting maximum number of CPUs to 4, total number of available CPUs is 4
13:51:09 etcd2 | 2016-03-23 13:51:09.746453 W | etcdmain: no data-dir provided, using default data-dir ./infra2.etcd
13:51:09 etcd2 | 2016-03-23 13:51:09.746613 I | etcdmain: listening for peers on http://127.0.0.1:22380
13:51:09 etcd2 | 2016-03-23 13:51:09.746673 I | etcdmain: listening for client requests on http://127.0.0.1:22379
13:51:09 etcd3 | 2016-03-23 13:51:09.749232 I | etcdmain: etcd Version: 2.3.0-alpha.1+git
13:51:09 etcd3 | 2016-03-23 13:51:09.749340 I | etcdmain: Git SHA: c7a7dfd
13:51:09 etcd3 | 2016-03-23 13:51:09.749354 I | etcdmain: Go Version: go1.6rc2
13:51:09 etcd3 | 2016-03-23 13:51:09.749361 I | etcdmain: Go OS/Arch: darwin/amd64
13:51:09 etcd3 | 2016-03-23 13:51:09.749372 I | etcdmain: setting maximum number of CPUs to 4, total number of available CPUs is 4
13:51:09 etcd3 | 2016-03-23 13:51:09.749384 W | etcdmain: no data-dir provided, using default data-dir ./infra3.etcd
13:51:09 etcd3 | 2016-03-23 13:51:09.749536 I | etcdmain: listening for peers on http://127.0.0.1:32380
13:51:09 etcd3 | 2016-03-23 13:51:09.749592 I | etcdmain: listening for client requests on http://127.0.0.1:32379
13:51:09 etcd2 | 2016-03-23 13:51:09.750614 I | etcdserver: name = infra2
13:51:09 etcd2 | 2016-03-23 13:51:09.750631 I | etcdserver: data dir = infra2.etcd
13:51:09 etcd2 | 2016-03-23 13:51:09.750638 I | etcdserver: member dir = infra2.etcd/member
13:51:09 etcd2 | 2016-03-23 13:51:09.750644 I | etcdserver: heartbeat = 100ms
13:51:09 etcd2 | 2016-03-23 13:51:09.750649 I | etcdserver: election = 1000ms
13:51:09 etcd2 | 2016-03-23 13:51:09.750654 I | etcdserver: snapshot count = 10000
13:51:09 etcd2 | 2016-03-23 13:51:09.750666 I | etcdserver: advertise client URLs = http://127.0.0.1:22379
13:51:09 etcd2 | 2016-03-23 13:51:09.750680 I | etcdserver: initial advertise peer URLs = http://127.0.0.1:22380
13:51:09 etcd2 | 2016-03-23 13:51:09.750709 I | etcdserver: initial cluster = infra1=http://127.0.0.1:12380,infra2=http://127.0.0.1:22380,infra3=http://127.0.0.1:32380
13:51:09 etcd2 | 2016-03-23 13:51:09.751325 I | etcdserver: starting member 91bc3c398fb3c146 in cluster ef37ad9dc622a7c4
13:51:09 etcd2 | 2016-03-23 13:51:09.751374 I | raft: 91bc3c398fb3c146 became follower at term 0
13:51:09 etcd2 | 2016-03-23 13:51:09.751407 I | raft: newRaft 91bc3c398fb3c146 [peers: [], term: 0, commit: 0, applied: 0, lastindex: 0, lastterm: 0]
13:51:09 etcd2 | 2016-03-23 13:51:09.751414 I | raft: 91bc3c398fb3c146 became follower at term 1
13:51:09 proxy | 2016-03-23 13:51:09.764839 I | etcdmain: etcd Version: 2.3.0-alpha.1+git
13:51:09 proxy | 2016-03-23 13:51:09.764969 I | etcdmain: Git SHA: c7a7dfd
13:51:09 proxy | 2016-03-23 13:51:09.764976 I | etcdmain: Go Version: go1.6rc2
13:51:09 proxy | 2016-03-23 13:51:09.764982 I | etcdmain: Go OS/Arch: darwin/amd64
13:51:09 proxy | 2016-03-23 13:51:09.764992 I | etcdmain: setting maximum number of CPUs to 4, total number of available CPUs is 4
13:51:09 proxy | 2016-03-23 13:51:09.765004 W | etcdmain: no data-dir provided, using default data-dir ./infra-proxy1.etcd
13:51:09 proxy | 2016-03-23 13:51:09.765205 I | etcdmain: proxy: using peer urls [http://127.0.0.1:12380 http://127.0.0.1:22380 http://127.0.0.1:32380] 
13:51:09 proxy | 2016-03-23 13:51:09.765977 W | etcdserver: could not get cluster response from http://127.0.0.1:12380: Get http://127.0.0.1:12380/members: dial tcp 127.0.0.1:12380: getsockopt: connection refused
13:51:09 etcd1 | 2016-03-23 13:51:09.773168 I | etcdmain: etcd Version: 2.3.0-alpha.1+git
13:51:09 etcd1 | 2016-03-23 13:51:09.774246 I | etcdmain: Git SHA: c7a7dfd
13:51:09 etcd1 | 2016-03-23 13:51:09.774259 I | etcdmain: Go Version: go1.6rc2
13:51:09 etcd1 | 2016-03-23 13:51:09.774267 I | etcdmain: Go OS/Arch: darwin/amd64
13:51:09 etcd1 | 2016-03-23 13:51:09.774283 I | etcdmain: setting maximum number of CPUs to 4, total number of available CPUs is 4
13:51:09 etcd1 | 2016-03-23 13:51:09.774300 W | etcdmain: no data-dir provided, using default data-dir ./infra1.etcd
13:51:09 etcd1 | 2016-03-23 13:51:09.774652 I | etcdmain: listening for peers on http://127.0.0.1:12380
13:51:09 etcd1 | 2016-03-23 13:51:09.774724 I | etcdmain: listening for client requests on http://127.0.0.1:12379
13:51:09 etcd2 | 2016-03-23 13:51:09.778444 I | etcdserver: starting server... [version: 2.3.0-alpha.1+git, cluster version: to_be_decided]
13:51:09 etcd2 | 2016-03-23 13:51:09.778818 I | etcdhttp: pprof is enabled under /debug/pprof
13:51:09 etcd2 | 2016-03-23 13:51:09.784269 E | etcdserver: cannot monitor file descriptor usage (cannot get FDUsage on darwin)
13:51:09 etcd1 | 2016-03-23 13:51:09.784713 I | etcdserver: name = infra1
13:51:09 etcd1 | 2016-03-23 13:51:09.784727 I | etcdserver: data dir = infra1.etcd
13:51:09 etcd1 | 2016-03-23 13:51:09.784733 I | etcdserver: member dir = infra1.etcd/member
13:51:09 etcd1 | 2016-03-23 13:51:09.784739 I | etcdserver: heartbeat = 100ms
13:51:09 etcd1 | 2016-03-23 13:51:09.784744 I | etcdserver: election = 1000ms
13:51:09 etcd1 | 2016-03-23 13:51:09.784749 I | etcdserver: snapshot count = 10000
13:51:09 etcd1 | 2016-03-23 13:51:09.784758 I | etcdserver: advertise client URLs = http://127.0.0.1:12379
13:51:09 etcd1 | 2016-03-23 13:51:09.784768 I | etcdserver: initial advertise peer URLs = http://127.0.0.1:12380
13:51:09 etcd1 | 2016-03-23 13:51:09.784787 I | etcdserver: initial cluster = infra1=http://127.0.0.1:12380,infra2=http://127.0.0.1:22380,infra3=http://127.0.0.1:32380
13:51:09 etcd1 | 2016-03-23 13:51:09.785342 I | etcdserver: starting member 8211f1d0f64f3269 in cluster ef37ad9dc622a7c4
13:51:09 etcd1 | 2016-03-23 13:51:09.785390 I | raft: 8211f1d0f64f3269 became follower at term 0
13:51:09 etcd1 | 2016-03-23 13:51:09.785412 I | raft: newRaft 8211f1d0f64f3269 [peers: [], term: 0, commit: 0, applied: 0, lastindex: 0, lastterm: 0]
13:51:09 etcd1 | 2016-03-23 13:51:09.785419 I | raft: 8211f1d0f64f3269 became follower at term 1
13:51:09 proxy | 2016-03-23 13:51:09.785341 I | etcdmain: proxy: listening for client requests on http://127.0.0.1:2379
13:51:09 etcd3 | 2016-03-23 13:51:09.784759 I | etcdserver: name = infra3
13:51:09 etcd3 | 2016-03-23 13:51:09.784772 I | etcdserver: data dir = infra3.etcd
13:51:09 etcd3 | 2016-03-23 13:51:09.784779 I | etcdserver: member dir = infra3.etcd/member
13:51:09 etcd3 | 2016-03-23 13:51:09.784784 I | etcdserver: heartbeat = 100ms
13:51:09 etcd3 | 2016-03-23 13:51:09.784789 I | etcdserver: election = 1000ms
13:51:09 etcd3 | 2016-03-23 13:51:09.784793 I | etcdserver: snapshot count = 10000
13:51:09 etcd3 | 2016-03-23 13:51:09.784807 I | etcdserver: advertise client URLs = http://127.0.0.1:32379
13:51:09 etcd3 | 2016-03-23 13:51:09.784816 I | etcdserver: initial advertise peer URLs = http://127.0.0.1:32380
13:51:09 etcd3 | 2016-03-23 13:51:09.784831 I | etcdserver: initial cluster = infra1=http://127.0.0.1:12380,infra2=http://127.0.0.1:22380,infra3=http://127.0.0.1:32380
13:51:09 etcd3 | 2016-03-23 13:51:09.785342 I | etcdserver: starting member fd422379fda50e48 in cluster ef37ad9dc622a7c4
13:51:09 etcd3 | 2016-03-23 13:51:09.785390 I | raft: fd422379fda50e48 became follower at term 0
13:51:09 etcd3 | 2016-03-23 13:51:09.785412 I | raft: newRaft fd422379fda50e48 [peers: [], term: 0, commit: 0, applied: 0, lastindex: 0, lastterm: 0]
13:51:09 etcd3 | 2016-03-23 13:51:09.785419 I | raft: fd422379fda50e48 became follower at term 1
13:51:09 etcd2 | 2016-03-23 13:51:09.784761 N | etcdserver: added member 8211f1d0f64f3269 [http://127.0.0.1:12380] to cluster ef37ad9dc622a7c4
13:51:09 etcd2 | 2016-03-23 13:51:09.784852 N | etcdserver: added local member 91bc3c398fb3c146 [http://127.0.0.1:22380] to cluster ef37ad9dc622a7c4
13:51:09 etcd2 | 2016-03-23 13:51:09.784921 N | etcdserver: added member fd422379fda50e48 [http://127.0.0.1:32380] to cluster ef37ad9dc622a7c4
13:51:09 etcd3 | 2016-03-23 13:51:09.794952 I | etcdserver: starting server... [version: 2.3.0-alpha.1+git, cluster version: to_be_decided]
13:51:09 etcd3 | 2016-03-23 13:51:09.795291 I | etcdhttp: pprof is enabled under /debug/pprof
13:51:09 etcd3 | 2016-03-23 13:51:09.797215 N | etcdserver: added member 8211f1d0f64f3269 [http://127.0.0.1:12380] to cluster ef37ad9dc622a7c4
13:51:09 etcd3 | 2016-03-23 13:51:09.797325 N | etcdserver: added member 91bc3c398fb3c146 [http://127.0.0.1:22380] to cluster ef37ad9dc622a7c4
13:51:09 etcd3 | 2016-03-23 13:51:09.799054 N | etcdserver: added local member fd422379fda50e48 [http://127.0.0.1:32380] to cluster ef37ad9dc622a7c4
13:51:09 etcd3 | 2016-03-23 13:51:09.801079 I | rafthttp: the connection with 91bc3c398fb3c146 became active
13:51:09 etcd2 | 2016-03-23 13:51:09.801258 I | rafthttp: the connection with fd422379fda50e48 became active
13:51:09 etcd3 | 2016-03-23 13:51:09.801168 E | etcdserver: cannot monitor file descriptor usage (cannot get FDUsage on darwin)
13:51:09 etcd1 | 2016-03-23 13:51:09.808323 I | etcdserver: starting server... [version: 2.3.0-alpha.1+git, cluster version: to_be_decided]
13:51:09 etcd1 | 2016-03-23 13:51:09.808635 I | etcdhttp: pprof is enabled under /debug/pprof
13:51:09 etcd1 | 2016-03-23 13:51:09.809104 I | rafthttp: the connection with fd422379fda50e48 became active
13:51:09 etcd1 | 2016-03-23 13:51:09.809371 I | rafthttp: the connection with 91bc3c398fb3c146 became active
13:51:09 etcd2 | 2016-03-23 13:51:09.809532 I | rafthttp: the connection with 8211f1d0f64f3269 became active
13:51:09 etcd3 | 2016-03-23 13:51:09.809586 I | rafthttp: the connection with 8211f1d0f64f3269 became active
13:51:09 etcd1 | 2016-03-23 13:51:09.811123 E | etcdserver: cannot monitor file descriptor usage (cannot get FDUsage on darwin)
13:51:09 etcd1 | 2016-03-23 13:51:09.813675 N | etcdserver: added local member 8211f1d0f64f3269 [http://127.0.0.1:12380] to cluster ef37ad9dc622a7c4
13:51:09 etcd1 | 2016-03-23 13:51:09.813827 N | etcdserver: added member 91bc3c398fb3c146 [http://127.0.0.1:22380] to cluster ef37ad9dc622a7c4
13:51:09 etcd1 | 2016-03-23 13:51:09.813962 N | etcdserver: added member fd422379fda50e48 [http://127.0.0.1:32380] to cluster ef37ad9dc622a7c4
13:51:10 etcd3 | 2016-03-23 13:51:10.187094 I | raft: fd422379fda50e48 is starting a new election at term 1
13:51:10 etcd3 | 2016-03-23 13:51:10.187215 I | raft: fd422379fda50e48 became candidate at term 2
13:51:10 etcd3 | 2016-03-23 13:51:10.187227 I | raft: fd422379fda50e48 received vote from fd422379fda50e48 at term 2
13:51:10 etcd3 | 2016-03-23 13:51:10.187239 I | raft: fd422379fda50e48 [logterm: 1, index: 3] sent vote request to 8211f1d0f64f3269 at term 2
13:51:10 etcd3 | 2016-03-23 13:51:10.187249 I | raft: fd422379fda50e48 [logterm: 1, index: 3] sent vote request to 91bc3c398fb3c146 at term 2
13:51:10 etcd2 | 2016-03-23 13:51:10.188224 I | raft: 91bc3c398fb3c146 [term: 1] received a MsgVote message with higher term from fd422379fda50e48 [term: 2]
13:51:10 etcd2 | 2016-03-23 13:51:10.188254 I | raft: 91bc3c398fb3c146 became follower at term 2
13:51:10 etcd2 | 2016-03-23 13:51:10.188319 I | raft: 91bc3c398fb3c146 [logterm: 1, index: 3, vote: 0] voted for fd422379fda50e48 [logterm: 1, index: 3] at term 2
13:51:10 etcd1 | 2016-03-23 13:51:10.188230 I | raft: 8211f1d0f64f3269 [term: 1] received a MsgVote message with higher term from fd422379fda50e48 [term: 2]
13:51:10 etcd1 | 2016-03-23 13:51:10.188262 I | raft: 8211f1d0f64f3269 became follower at term 2
13:51:10 etcd1 | 2016-03-23 13:51:10.188329 I | raft: 8211f1d0f64f3269 [logterm: 1, index: 3, vote: 0] voted for fd422379fda50e48 [logterm: 1, index: 3] at term 2
13:51:10 etcd3 | 2016-03-23 13:51:10.188774 I | raft: fd422379fda50e48 received vote from 8211f1d0f64f3269 at term 2
13:51:10 etcd3 | 2016-03-23 13:51:10.188798 I | raft: fd422379fda50e48 [quorum:2] has received 2 votes and 0 vote rejections
13:51:10 etcd3 | 2016-03-23 13:51:10.188836 I | raft: fd422379fda50e48 became leader at term 2
13:51:10 etcd3 | 2016-03-23 13:51:10.188868 I | raft: raft.node: fd422379fda50e48 elected leader fd422379fda50e48 at term 2
13:51:10 etcd1 | 2016-03-23 13:51:10.189121 I | raft: raft.node: 8211f1d0f64f3269 elected leader fd422379fda50e48 at term 2
13:51:10 etcd2 | 2016-03-23 13:51:10.189364 I | raft: raft.node: 91bc3c398fb3c146 elected leader fd422379fda50e48 at term 2
13:51:10 etcd3 | 2016-03-23 13:51:10.190826 I | etcdserver: published {Name:infra3 ClientURLs:[http://127.0.0.1:32379]} to cluster ef37ad9dc622a7c4
13:51:10 etcd1 | 2016-03-23 13:51:10.191134 I | etcdserver: published {Name:infra1 ClientURLs:[http://127.0.0.1:12379]} to cluster ef37ad9dc622a7c4
13:51:10 etcd2 | 2016-03-23 13:51:10.191299 I | etcdserver: published {Name:infra2 ClientURLs:[http://127.0.0.1:22379]} to cluster ef37ad9dc622a7c4
13:51:10 etcd3 | 2016-03-23 13:51:10.191529 I | etcdserver: setting up the initial cluster version to 2.3
13:51:10 etcd3 | 2016-03-23 13:51:10.191959 N | etcdserver: set the initial cluster version to 2.3
13:51:10 etcd1 | 2016-03-23 13:51:10.192055 N | etcdserver: set the initial cluster version to 2.3
13:51:10 etcd2 | 2016-03-23 13:51:10.192095 N | etcdserver: set the initial cluster version to 2.3
13:51:10 proxy | 2016-03-23 13:51:10.787589 I | proxy: endpoints found ["http://127.0.0.1:22379" "http://127.0.0.1:32379" "http://127.0.0.1:12379"]
```

```
➜  etcd git:(master) ✗ ./bin/etcdctl member list 
8211f1d0f64f3269: name=infra1 peerURLs=http://127.0.0.1:12380 clientURLs=http://127.0.0.1:12379
91bc3c398fb3c146: name=infra2 peerURLs=http://127.0.0.1:22380 clientURLs=http://127.0.0.1:22379
fd422379fda50e48: name=infra3 peerURLs=http://127.0.0.1:32380 clientURLs=http://127.0.0.1:32379
```

```
➜  etcd git:(master) ✗ ./bin/etcdctlv3 member list 
2016/03/23 13:56:52 grpc: Conn.resetTransport failed to create client transport: connection error: desc = "transport: dial tcp 127.0.0.1:2378: getsockopt: connection refused"; Reconnecting to "127.0.0.1:2378"
2016/03/23 13:56:53 grpc: Conn.resetTransport failed to create client transport: connection error: desc = "transport: dial tcp 127.0.0.1:2378: getsockopt: connection refused"; Reconnecting to "127.0.0.1:2378"
2016/03/23 13:56:56 grpc: Conn.resetTransport failed to create client transport: connection error: desc = "transport: dial tcp 127.0.0.1:2378: getsockopt: connection refused"; Reconnecting to "127.0.0.1:2378"
2016/03/23 13:56:59 grpc: Conn.resetTransport failed to create client transport: connection error: desc = "transport: dial tcp 127.0.0.1:2378: getsockopt: connection refused"; Reconnecting to "127.0.0.1:2378"
^C
```

```
➜  etcd git:(master) ✗ ./bin/etcdctlv3 --endpoints=[127.0.0.1:12379,127.0.0.1:22379,127.0.0.1:32379] member list 
zsh: no matches found: --endpoints=[127.0.0.1:12379,127.0.0.1:22379,127.0.0.1:32379]
```

xiang90 commented:
@mqliang You need to run `goreman start -f V3DemoProcfile start`, not `goreman start`. check my previous comment.

mqliang commented:
Aha, have a try. Unfortunately, there seems a bug in goreman or something of V3DemoProcfile goes wrong.

```
➜  etcd git:(master) ✗ goreman start -f V3DemoProcfile start
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xb code=0x1 addr=0x28 pc=0x48a4]

goroutine 1 [running]:
main.startProc(0x7fff5fbffbbe, 0x2, 0x0, 0x0)
	/Users/mqliang/Documents/NetEase/Etcd/src/github.com/mattn/goreman/proc.go:60 +0x1b4
main.startProcs(0x0, 0x0)
	/Users/mqliang/Documents/NetEase/Etcd/src/github.com/mattn/goreman/proc.go:86 +0xda
main.start(0xc82011c1e0, 0x0, 0x0)
	/Users/mqliang/Documents/NetEase/Etcd/src/github.com/mattn/goreman/goreman.go:166 +0x23e
main.main()
	/Users/mqliang/Documents/NetEase/Etcd/src/github.com/mattn/goreman/goreman.go:202 +0x730
```

Install goreman by `go get github.com/mattn/goreman`

xiang90 commented:
goreman -f V3DemoProcfile start?

mqliang commented:
@xiang90 Everything works fine. Great thanks for your guiding! 

"--command-timeout" flag could work now, PR is ready for review. PTAL.

xiang90 commented:
LGTM. defer to @heyitsanthony (shall we wait for the vendor merge? or we shall merge this first?)

heyitsanthony commented:
Would it be better to replace code like:
```go
ctx, cancel := commandCtx(cmd)
whatever()
cancel()
```
with 
```go
ctx, _ := commandCtx(cmd)
whatever()
```
?

All the `cancel()`s seem unnecessary to me. lgtm aside from that

mqliang commented:
@heyitsanthony 
Copy from godocs of `WithTimeout`

```
// WithTimeout returns WithDeadline(parent, time.Now().Add(timeout)).
//
// Canceling this context releases resources associated with it, so code should
// call cancel as soon as the operations running in this Context complete:
//
// 	func slowOperationWithTimeout(ctx context.Context) (Result, error) {
// 		ctx, cancel := context.WithTimeout(ctx, 100*time.Millisecond)
// 		defer cancel()  // releases resources if slowOperation completes before timeout elapses
// 		return slowOperation(ctx)
// 	}
func WithTimeout(parent Context, timeout time.Duration) (Context, CancelFunc) {
	return WithDeadline(parent, time.Now().Add(timeout))
}
```

Why do you think all the `cancel()`s unnecessary.

heyitsanthony commented:
@mqliang I think they're unnecessary because etcdctl exits shortly after completing the command...

fix the merge conflicts and lgtm.

mqliang commented:
Updated. PTAL. @heyitsanthony keep all the `cancel()` stuff since it's recommended by godocs, if you insist on removing them, I can rework. I am fine with that.

heyitsanthony commented:
lgtm
